### ISSUE TITLE
Issues with Mutability

### ISSUE THREAD

**Issue by JKTKops** [JKTKops/Thcheme#1]
Scheme strings and vectors are supposed to be mutable. To this end, thcheme provides `string-set!` and `vector-set!` macros. This is all well and good, except that those are supposed to be functions. Why are they macros currently? Who knows! But there's a bigger problem: Haskell Strings and Vectors aren't mutable. `string-set!` can only mutate the string relative to a symbol its defined as. Therefore, strings should actually be `IORef String`s, and appropriate work to access and change them is necessary.

The exact problem is something like this:
```scheme
$ thcheme
loaded: stdlib.thm
>>> (define str "foo")
"foo"
>>> (string-set! str 0 #\z)
"zoo"
>>> str
"zoo"
>>> (string-set! (id str) 0 #\f)
Error: Invalid type: expected string, found (id str)
Stack trace:
1; (string-set! (id str) 0 #\f)
```
`(id str)` should've been evaluated to a _mutable_ string, and then adjusted, and evaluating `str` again should return "foo". Currently, this isn't possible, because `(id str)` evaluated to the `LispVal` `LispString String` - no IO, so no mutation.

Vectors are hopefully a little bit easier to fix, using an IO-mutable vector instead of just a vector. I'll attempt to attack this now.

**Comment by JKTKops** [comment]
Vectors aren't really easier to fix. Lots of functions right now rely on LispVal being IO-free, so I'll need to go through and put anything that can inspect vectors (and strings) into IO.

**Comment by JKTKops** [comment]
It turned out there was an interesting thing that was even _more_ wrong.

`eq?` should be able to detect when two lists are the same object. It couldn't, because `(eq? (cdr lst) (cdr lst))` would test if two distinct (but structurally equivalent) Haskell `List` objects were the same object, and they aren't. I wouldn't have noticed this before, but this was exposed as I started making things properly mutable. This included pairs. Pairs have been properly mutable for a few days now and I've also corrected the implementations of `eq?`, `eqv?`, and `equal?` (the latter using the algorithm presented in Efficient Nondestructive Equality Checking for Trees and Graphs, R. Dybvig et al). All pairs are now truly pairs. Mutable pairs contain `Ref`s and immutable pairs contain values directly.

Properly mutable vectors were added with 05b6e67. Furthermore this commit switched from `array` to `vector` (I was much less familiar with the ecosystem in 2018).

Mutable strings are still outstanding.
